session/nontransactional.go

```diff
@@ -759,37 +768,138 @@ func buildShardJobs(ctx context.Context, stmt *ast.NonTransactionalDMLStmt, se S
 	return jobs, nil
 }
 
+// Int64Slice attaches the methods of Interface to []int64, sorting in increasing order.
+type Int64Slice []int64
+
+func (x Int64Slice) Len() int           { return len(x) }
+func (x Int64Slice) Less(i, j int) bool { return x[i] < x[j] }
+func (x Int64Slice) Swap(i, j int)      { x[i], x[j] = x[j], x[i] }
+
+func buildShardJobsUnordered(ctx context.Context, stmt *ast.NonTransactionalDMLStmt, se Session,
+	selectSQL string, memTracker *memory.Tracker) ([]job, error) {
+	// A NT-DML is not a SELECT. We ignore the SelectLimit for selectSQL so that it can read all values.
+	originalSelectLimit := se.GetSessionVars().SelectLimit
+	se.GetSessionVars().SelectLimit = math.MaxUint64
+	// NT-DML is a write operation, and should not be affected by read_staleness that is supposed to affect only SELECT.
+	rss, err := se.Execute(ctx, selectSQL)
+	se.GetSessionVars().SelectLimit = originalSelectLimit
+
+	if err != nil {
+		return nil, err
+	}
+	if len(rss) != 1 {
+		return nil, errors.Errorf("Non-transactional DML, expecting 1 record set, but got %d", len(rss))
+	}
+	rs := rss[0]
+	defer func() {
+		_ = rs.Close()
+	}()
+
+	batchSize := int(stmt.Limit)
+	if batchSize <= 0 {
+		return nil, errors.New("Non-transactional DML, batch size should be positive")
+	}
+
+	rowIDs := make([]int64, 0, 2<<20)
+
+	chk := rs.NewChunk(nil)
+	for {
+		err = rs.Next(ctx, chk)
+		if err != nil {
+			return nil, err
+		}
+		// last chunk
+		if chk.NumRows() == 0 {
+			break
+		}
+		iter := chunk.NewIterator4Chunk(chk)
+		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+			current := row.GetDatum(0, &rs.Fields()[0].Column.FieldType)
+			rowIDs = append(rowIDs, current.GetInt64())
```

Comment:
Storing all rowIDs can still result in uncontrolled memory usage.

Reply:
The available memory is enough for now.